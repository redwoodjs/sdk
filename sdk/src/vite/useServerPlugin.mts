export { transformServerFunctions } from "./transformServerFunctions.mjs";
