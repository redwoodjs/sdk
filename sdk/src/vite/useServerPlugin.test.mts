import { describe, expect, it } from "vitest";
import { transformServerFunctions } from "./useServerPlugin.mjs";

describe("useServerPlugin", () => {
  it("determines react server directive", () => {
    expect(
      transformServerFunctions(`\
"use server";
            `),
    ).toMatchInlineSnapshot(`
      "
                  "
    `);

    expect(
      transformServerFunctions(`\
// These are not server functions
            `),
    ).toMatchInlineSnapshot(`undefined`);

    expect(
      transformServerFunctions(`\
// Comment
"use server";
              `),
    ).toMatchInlineSnapshot(`
      "// Comment

                    "
    `);

    expect(
      transformServerFunctions(`\
// Multi-line
// Comment
"use server";
                `),
    ).toMatchInlineSnapshot(`
      "// Multi-line
      // Comment

                      "
    `);

    expect(
      transformServerFunctions(`\
/* Giant
 * Comment
 * Block
 */

"use server";
                  `),
    ).toMatchInlineSnapshot(`
      "/* Giant
       * Comment
       * Block
       */


                        "
    `);
  });

  it("supports default exports", () => {
    expect(
      transformServerFunctions(`\
  "use server";

  export default function execute() {
    return 1 + 1;
  }
              `),
    ).toMatchInlineSnapshot(`
      "  

        export function execute() {
          return 1 + 1;
        }

      export default execute;
                    "
    `);
  });
});
