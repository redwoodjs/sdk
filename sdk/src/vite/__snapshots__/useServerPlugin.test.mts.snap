// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useServerPlugin > TRANSFORMS > ARROW_FUNCTION_EXPORT_CODE > CLIENT 1`] = `
"import { createServerReference } from "rwsdk/client";

export let sum = createServerReference("/test.tsx", "sum");
"
`;

exports[`useServerPlugin > TRANSFORMS > ARROW_FUNCTION_EXPORT_CODE > WORKER 1`] = `
"import { registerServerReference } from "rwsdk/worker";

export const sum = () => {
  return 1 + 1
}
registerServerReference(sum, "/test.tsx", "sum")
"
`;

exports[`useServerPlugin > TRANSFORMS > ASYNC_FUNCTION_EXPORT_CODE > CLIENT 1`] = `
"import { createServerReference } from "rwsdk/client";

export let sum = createServerReference("/test.tsx", "sum");
"
`;

exports[`useServerPlugin > TRANSFORMS > ASYNC_FUNCTION_EXPORT_CODE > WORKER 1`] = `
"import { registerServerReference } from "rwsdk/worker";

export async function sum() {
  return 1 + 1
}
registerServerReference(sum, "/test.tsx", "sum")
"
`;

exports[`useServerPlugin > TRANSFORMS > COMMENT_BLOCK_CODE > CLIENT 1`] = `
"import { createServerReference } from "rwsdk/client";

export let sum = createServerReference("/test.tsx", "sum");
"
`;

exports[`useServerPlugin > TRANSFORMS > COMMENT_BLOCK_CODE > WORKER 1`] = `
"
/* Giant
 * Comment
 * Block
 */
import { registerServerReference } from "rwsdk/worker";

export function sum() {
  return 1 + 1
}
registerServerReference(sum, "/test.tsx", "sum")
"
`;

exports[`useServerPlugin > TRANSFORMS > COMMENT_CODE > CLIENT 1`] = `
"import { createServerReference } from "rwsdk/client";

export let sum = createServerReference("/test.tsx", "sum");
"
`;

exports[`useServerPlugin > TRANSFORMS > COMMENT_CODE > WORKER 1`] = `
"import { registerServerReference } from "rwsdk/worker";

// Comment


export function sum() {
  return 1 + 1;
}
registerServerReference(sum, "/test.tsx", "sum")
"
`;

exports[`useServerPlugin > TRANSFORMS > DEFAULT_AND_NAMED_EXPORTS_CODE > CLIENT 1`] = `
"import { createServerReference } from "rwsdk/client";

export let sum = createServerReference("/test.tsx", "sum");

export default createServerReference("/test.tsx", "default");
"
`;

exports[`useServerPlugin > TRANSFORMS > DEFAULT_AND_NAMED_EXPORTS_CODE > WORKER 1`] = `
"import { registerServerReference } from "rwsdk/worker";

export function sum() {
  return 1 + 1;
}

function __defaultServerFunction__() {
  return 1 + 2;
}

export default __defaultServerFunction__;
registerServerReference(__defaultServerFunction__, "/test.tsx", "default")
registerServerReference(sum, "/test.tsx", "sum")
"
`;

exports[`useServerPlugin > TRANSFORMS > DEFAULT_EXPORT_CODE > CLIENT 1`] = `
"import { createServerReference } from "rwsdk/client";

export let sum = createServerReference("/test.tsx", "sum");

export default createServerReference("/test.tsx", "default");
"
`;

exports[`useServerPlugin > TRANSFORMS > DEFAULT_EXPORT_CODE > WORKER 1`] = `
"import { registerServerReference } from "rwsdk/worker";

function __defaultServerFunction__() {
  return 1 + 1;
}

export default __defaultServerFunction__;
registerServerReference(__defaultServerFunction__, "/test.tsx", "default")
"
`;

exports[`useServerPlugin > TRANSFORMS > MULTI_LINE_COMMENT_CODE > CLIENT 1`] = `
"import { createServerReference } from "rwsdk/client";

export let sum = createServerReference("/test.tsx", "sum");
"
`;

exports[`useServerPlugin > TRANSFORMS > MULTI_LINE_COMMENT_CODE > WORKER 1`] = `
"import { registerServerReference } from "rwsdk/worker";

// Multi-line
// Comment


export function sum() {
  return 1 + 1
}
registerServerReference(sum, "/test.tsx", "sum")
"
`;

exports[`useServerPlugin > TRANSFORMS > NAMED_EXPORT_CODE > CLIENT 1`] = `
"import { createServerReference } from "rwsdk/client";

export let sum = createServerReference("/test.tsx", "sum");
"
`;

exports[`useServerPlugin > TRANSFORMS > NAMED_EXPORT_CODE > WORKER 1`] = `
"import { registerServerReference } from "rwsdk/worker";

export function sum() {
  return 1 + 1
}
registerServerReference(sum, "/test.tsx", "sum")
"
`;
